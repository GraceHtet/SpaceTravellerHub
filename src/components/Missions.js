import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions } from '../redux/missions/missionsSlice';
import styles from '../styles/Missions.module.css';

const Missions = () => {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);

  useEffect(() => {
    if (!missions.length) {
      dispatch(fetchMissions());
    }
  }, [dispatch, missions]);

  return (
    <div>
      <h1 className={styles.title}>Missions</h1>

      <table>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission) => (
            <tr key={mission.mission_id}>
              <td>{mission.mission_name}</td>
              <td>{mission.description}</td>
              <td>{mission.status}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};
export default Missions;
