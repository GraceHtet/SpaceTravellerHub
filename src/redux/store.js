import { confugureStore } from '@reduxjs/toolkit';

const store = configureStore({
  reducer: {},
});

export default store;
