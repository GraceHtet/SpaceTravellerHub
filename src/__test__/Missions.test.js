import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import TestRenderer, { act } from 'react-test-renderer';
import Missions from '../components/Missions';
import store from '../redux/store';

describe('Missions', () => {
  test('Render Missions Component', () => {
    const tree = TestRenderer.create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Join Mission Buttion', async () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    const joinButton = await screen.findAllByText('Join Mission');
    expect(joinButton).toHaveLength(10);
  });

  test('Clicking Leave Mission Button', async () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    act(() => {
      const joinButton = screen.getAllByText('Join Mission');
      expect(joinButton).toHaveLength(10);
      joinButton[0].click();
    });

    act(() => {
      const leaveButton = screen.getAllByText('Leave Mission');
      leaveButton[0].click();
    });

    const joinButton = await screen.findAllByText('Join Mission');
    expect(joinButton).toHaveLength(10);
  });

  test('Clicking Join Mission Button', async () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    act(() => {
      const joinButton = screen.getAllByText('Join Mission');
      joinButton[0].click();
    });

    const leaveButton = await screen.findAllByText('Leave Mission');
    expect(leaveButton).toHaveLength(1);
  });
});
