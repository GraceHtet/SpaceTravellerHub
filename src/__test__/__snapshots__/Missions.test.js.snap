// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Render Missions Component 1`] = `
<div>
  <h1
    className="title"
  >
    Missions
  </h1>
  <p
    className="loading"
  >
    Loading missions...
  </p>
  <table
    className="table"
  >
    <thead
      className="thead"
    >
      <tr
        className="tr"
      >
        <th
          className="th"
        >
          Mission
        </th>
        <th
          className="th"
        >
          Description
        </th>
        <th
          className="th status"
        >
          Status
        </th>
        <th
          className="th action"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
